README.md
# quiz_view

A new Flutter package project that helps create quizes and surveys. It has a quiz_view customizable widget.

## Screenshots

<img src="ss1.png" height="300em" /> <img src="ss2.png" height="300em" />

## Usage

[Example](https://github.com/JayTWWM/Quiz-View-Flutter/blob/master/example/example_app.dart)

To use this package :

* add the dependency to your [pubspec.yaml] file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    quiz_view:
```

### How to use

```dart
QuizView(
  image: Container(
    width: 150,
    height: 150,
    child: Image.network(
        "https://yt3.ggpht.com/a/AATXAJyPMywRmD62sfK-1CXjwF0YkvrvnmaaHzs4uw=s900-c-k-c0xffffffff-no-rj-mo"),
  ),
  showCorrect: true,
  tagBackgroundColor: Colors.deepOrange,
  tagColor: Colors.black,
  questionTag: "Question: 2",
  answerColor: Colors.white,
  answerBackgroundColor: Color.fromARGB(255, 255, 0, 111),
  questionColor: Colors.white,
  backgroundColor: Color.fromARGB(255, 111, 0, 255),
  width: 300,
  height: 600,
  question: "Which is the best framework for app development?",
  rightAnswer: "Flutter",
  wrongAnswers: ["Fluttor", "Flitter"],
  onRightAnswer: () => print("Right"),
  onWrongAnswer: () => print("Wrong"),
)
```

# License

    Copyright 2020 Jay Mehta

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

## Example

As time based...

``` dart
import 'package:flutter/material.dart';
import 'package:quiz_view/quiz_view.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Quiz View Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.deepOrange,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Quiz View Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: QuizView(
        image: Container(
          width: 150,
          height: 150,
          child: Image.network(
              "https://yt3.ggpht.com/a/AATXAJyPMywRmD62sfK-1CXjwF0YkvrvnmaaHzs4uw=s900-c-k-c0xffffffff-no-rj-mo"),
        ),
        showCorrect: true,
        tagBackgroundColor: Colors.deepOrange,
        tagColor: Colors.black,
        questionTag: "Question: 2",
        answerColor: Colors.white,
        answerBackgroundColor: Color.fromARGB(255, 255, 0, 111),
        questionColor: Colors.white,
        backgroundColor: Color.fromARGB(255, 111, 0, 255),
        width: 300,
        height: 600,
        question: "Which is the best framework for app development?",
        rightAnswer: "Flutter",
        wrongAnswers: ["Fluttor", "Flitter"],
        onRightAnswer: () => print("Right"),
        onWrongAnswer: () => print("Wrong"),
      )),
    );
  }
}

```


CHANGELOG.md
## [0.0.1] - Release

* A customizable quiz view widget.

## [0.0.2] - Release

* A customizable quiz view widget.

## [0.0.3] - Release

* A customizable quiz view widget.

## [0.0.4] - Release

* A customizable quiz view widget.

## [0.0.5] - Release

* A customizable quiz view widget.

## [1.0.0] - Release

* A customizable quiz view widget.

## [1.0.1] - Release

* A customizable quiz view widget.

## [1.0.2] - Release

* A customizable quiz view widget.

## [1.0.3] - Release

* A customizable quiz view widget.

## [1.0.4] - Release

* A customizable quiz view widget.

## [1.0.5] - Release

* A customizable quiz view widget.


lib/quiz_view.dart
library quiz_view;

import 'package:flutter/material.dart';
import 'dart:math';

class QuizView extends StatefulWidget {
  /// Boolean to show the correct answer after the quiz is answered
  final bool showCorrect;

  /// The question
  final String question;

  /// The questonTag (question number or id)
  final String questionTag;

  /// Color of question font
  final Color questionColor;

  /// Background color
  final Color backgroundColor;

  /// Image if any
  final Widget image;

  /// Width of the quiz view
  final double width;

  /// Height of the quiz view
  final double height;

  /// The right answer
  final String rightAnswer;

  /// The wrong answers
  final List<String> wrongAnswers;

  /// Question Tag background color
  final Color tagBackgroundColor;

  /// Question Tag font color
  final Color tagColor;

  /// Answer font color
  final Color answerColor;

  /// Answer background color
  final Color answerBackgroundColor;

  /// This function is executed if the answer is right
  final void Function() onRightAnswer;

  /// This function is executed if the answer is wrong
  final void Function() onWrongAnswer;

  QuizView(
      {this.showCorrect = true,
      this.questionTag,
      @required this.question,
      this.questionColor = Colors.black,
      this.backgroundColor = Colors.white,
      this.image,
      @required this.height,
      @required this.width,
      @required this.rightAnswer,
      @required this.wrongAnswers,
      this.tagColor = Colors.black,
      this.tagBackgroundColor = Colors.white,
      this.answerColor = Colors.black,
      this.answerBackgroundColor = Colors.white,
      @required this.onRightAnswer,
      @required this.onWrongAnswer});

  _QuizViewState createState() => _QuizViewState();
}

class _QuizViewState extends State<QuizView> {
  _QuizViewState();

  bool isTapped = false;
  static Random _random = new Random();
  int answerIndex;

  @override
  Widget build(BuildContext context) {
    if (!isTapped) {
      answerIndex = _random.nextInt(widget.wrongAnswers.length + 1);
    }

    Column answerColumn = Column(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [],
    );

    for (String i in widget.wrongAnswers) {
      answerColumn.children.add(Container(
        width: MediaQuery.of(context).size.width,
        padding: EdgeInsets.all(15),
        child: ElevatedButton(
          style: ButtonStyle(
            backgroundColor: MaterialStateProperty.resolveWith(
                (states) => widget.answerBackgroundColor),
            padding: MaterialStateProperty.resolveWith(
                (states) => EdgeInsets.fromLTRB(20, 10, 20, 10)),
            shape:
                MaterialStateProperty.resolveWith((states) => StadiumBorder()),
            textStyle: MaterialStateProperty.resolveWith(
                (states) => TextStyle(color: Colors.white)),
          ),
          child: Center(
            child: Text(
              i,
              style: TextStyle(
                  color: widget.answerColor,
                  fontSize: widget.width > widget.height
                      ? widget.width / 25
                      : widget.height / 25),
            ),
          ),
          onPressed: () {
            if (!isTapped) {
              widget.onWrongAnswer();
              if (widget.showCorrect) {
                setState(() {
                  isTapped = !isTapped;
                });
              }
            }
          },
        ),
      ));
    }
    answerColumn.children.insert(
        answerIndex,
        Container(
          width: MediaQuery.of(context).size.width,
          padding: EdgeInsets.all(15),
          child: ElevatedButton(
            style: ButtonStyle(
              backgroundColor: MaterialStateProperty.resolveWith((states) =>
                  isTapped ? Colors.green : widget.answerBackgroundColor),
              padding: MaterialStateProperty.resolveWith(
                  (states) => EdgeInsets.fromLTRB(20, 10, 20, 10)),
              shape: MaterialStateProperty.resolveWith(
                  (states) => StadiumBorder()),
              textStyle: MaterialStateProperty.resolveWith(
                  (states) => TextStyle(color: Colors.white)),
            ),
            child: Center(
              child: Text(
                widget.rightAnswer,
                style: TextStyle(
                    color: widget.answerColor,
                    fontSize: widget.width > widget.height
                        ? widget.width / 25
                        : widget.height / 25),
              ),
            ),
            onPressed: () {
              if (!isTapped) {
                widget.onRightAnswer();
                if (widget.showCorrect) {
                  setState(() {
                    isTapped = !isTapped;
                  });
                }
              }
            },
          ),
        ));

    return Container(
        decoration: BoxDecoration(
            color: widget.backgroundColor,
            borderRadius: BorderRadius.all(Radius.circular(
                widget.height > widget.width
                    ? widget.height / 20
                    : widget.width / 20)),
            border: Border.all(
                color: Colors.black, width: 1.5, style: BorderStyle.solid)),
        width: widget.width,
        height: widget.height,
        child: widget.questionTag != null
            ? Stack(
                children: [
                  Positioned(
                    top: 0,
                    left: 0,
                    child: Container(
                      padding: EdgeInsets.all(10),
                      decoration: BoxDecoration(
                          color: widget.tagBackgroundColor,
                          borderRadius: BorderRadius.only(
                            topLeft: Radius.circular(
                                widget.height > widget.width
                                    ? widget.height / 20
                                    : widget.width / 20),
                            bottomRight: Radius.circular(
                                widget.height > widget.width
                                    ? widget.height / 20
                                    : widget.width / 20),
                          )),
                      child: Text(
                        widget.questionTag,
                        style: TextStyle(
                            color: widget.tagColor,
                            fontWeight: FontWeight.bold,
                            fontSize: widget.width > widget.height
                                ? widget.width / 20
                                : widget.height / 20),
                      ),
                    ),
                  ),
                  Container(
                    margin: EdgeInsets.fromLTRB(
                        0,
                        widget.width > widget.height
                            ? widget.width / 10
                            : widget.height / 10,
                        0,
                        0),
                    child: SingleChildScrollView(
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                        crossAxisAlignment: CrossAxisAlignment.center,
                        children: widget.image != null
                            ? [
                                Container(
                                  padding: EdgeInsets.all(20),
                                  child: RichText(
                                      text: TextSpan(
                                    text: widget.question,
                                    style: TextStyle(
                                        color: widget.questionColor,
                                        fontWeight: FontWeight.bold,
                                        fontSize: widget.width > widget.height
                                            ? widget.width / 20
                                            : widget.height / 20),
                                  )),
                                ),
                                widget.image,
                                answerColumn
                              ]
                            : [
                                Container(
                                  padding: EdgeInsets.all(20),
                                  child: RichText(
                                    text: TextSpan(
                                        text: widget.question,
                                        style: TextStyle(
                                            color: widget.questionColor,
                                            fontWeight: FontWeight.bold,
                                            fontSize:
                                                widget.width > widget.height
                                                    ? widget.width / 20
                                                    : widget.height / 20)),
                                  ),
                                ),
                                answerColumn
                              ],
                      ),
                    ),
                  ),
                ],
              )
            : SingleChildScrollView(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: widget.image != null
                      ? [
                          Container(
                            padding: EdgeInsets.all(20),
                            child: RichText(
                                text: TextSpan(
                              text: widget.question,
                              style: TextStyle(
                                  color: widget.questionColor,
                                  fontWeight: FontWeight.bold,
                                  fontSize: widget.width > widget.height
                                      ? widget.width / 20
                                      : widget.height / 20),
                            )),
                          ),
                          widget.image,
                          answerColumn
                        ]
                      : [
                          Container(
                            padding: EdgeInsets.all(20),
                            child: RichText(
                              text: TextSpan(
                                  text: widget.question,
                                  style: TextStyle(
                                      color: widget.questionColor,
                                      fontWeight: FontWeight.bold,
                                      fontSize: widget.width > widget.height
                                          ? widget.width / 20
                                          : widget.height / 20)),
                            ),
                          ),
                          answerColumn
                        ],
                ),
              ));
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:quiz_view/quiz_view.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Quiz View Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.deepOrange,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Quiz View Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: QuizView(
        image: Container(
          width: 150,
          height: 150,
          child: Image.network(
              "https://yt3.ggpht.com/a/AATXAJyPMywRmD62sfK-1CXjwF0YkvrvnmaaHzs4uw=s900-c-k-c0xffffffff-no-rj-mo"),
        ),
        showCorrect: true,
        tagBackgroundColor: Colors.deepOrange,
        tagColor: Colors.black,
        questionTag: "Question: 2",
        answerColor: Colors.white,
        answerBackgroundColor: Color.fromARGB(255, 255, 0, 111),
        questionColor: Colors.white,
        backgroundColor: Color.fromARGB(255, 111, 0, 255),
        width: 300,
        height: 600,
        question: "Which is the best framework for app development?",
        rightAnswer: "Flutter",
        wrongAnswers: ["Fluttor", "Flitter"],
        onRightAnswer: () => print("Right"),
        onWrongAnswer: () => print("Wrong"),
      )),
    );
  }
}
